pkg/local-gadget-manager/local-gadget-manager.go

```diff
@@ -0,0 +1,309 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package localgadgetmanager
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/container-collection"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadget-collection"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/containerutils"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/stream"
+	"github.com/kinvolk/inspektor-gadget/pkg/runcfanotify"
+
+	dockertypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type LocalGadgetManager struct {
+	containercollection.ContainerCollection
+
+	traceFactories map[string]gadgets.TraceFactory
+
+	// tracers by name
+	tracers map[string]tracer
+
+	runcNotifier *runcfanotify.RuncNotifier
+
+	dockercli *client.Client
+}
+
+type tracer struct {
+	gadget        string
+	name          string
+	factory       gadgets.TraceFactory
+	traceResource *gadgetv1alpha1.Trace
+	gadgetStream  *stream.GadgetStream
+}
+
+func (l *LocalGadgetManager) ListGadgets() []string {
+	gadgets := []string{}
+	for name := range l.traceFactories {
+		gadgets = append(gadgets, name)
+	}
+	sort.Strings(gadgets)
+	return gadgets
+}
+
+func (l *LocalGadgetManager) ListOperations(gadget string) []string {
+	return []string{"start", "stop", "generate"}
+}
+
+func (l *LocalGadgetManager) ListTraces() []string {
+	traces := []string{}
+	for name := range l.tracers {
+		traces = append(traces, name)
+	}
+	sort.Strings(traces)
+	return traces
+}
+
+func (l *LocalGadgetManager) ListContainers() []string {
+	containers := []string{}
+	l.ContainerRange(func(c pb.ContainerDefinition) {
+		containers = append(containers, c.Id)
+	})
+	sort.Strings(containers)
+	return containers
+}
+
+func (l *LocalGadgetManager) AddTracer(gadget, name, containerFilter string) error {
+	factory, ok := l.traceFactories[gadget]
+	if !ok {
+		return fmt.Errorf("unknown gadget")
+	}
+	_, ok = l.tracers[name]
+	if ok {
+		fmt.Printf("Error: trace already exists")
+		return nil
+	}
+
+	traceResource := &gadgetv1alpha1.Trace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: "gadget",
+		},
+		Spec: gadgetv1alpha1.TraceSpec{
+			Node:   "local",
+			Gadget: gadget,
+			Filter: &gadgetv1alpha1.ContainerFilter{
+				Namespace:     "local",
+				Podname:       containerFilter,
+				ContainerName: "",
+				Labels:        map[string]string{},
+			},
+			RunMode:    "Manual",
+			OutputMode: "Stream",
```

Comment:
Why always "Stream"? What about gadget that don't support it?

Reply:
I was wondering the same and after a quick check, I think `RunMode` is not being used and `OutputMode` is used just by seccomp gadget which is currently the only gadget to support multiple output modes. Therefore, I think we should include this parameter only for seccomp and add a comment explaining all this. WDYT?